### DISCUSSION TITLE
Define post-validation type of a field

### DISCUSSION THREAD

**Discussion by caniko** [pydantic/pydantic#7511] [Feature Request]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

I have a field with type `Union[A, B, C]`, after validation the field will have type `C`. Is it possible to reflect this in the typing of the field. This can also be considered for annotated types that does some validation intrinsic to the type itself, and returns a pre-dermined type.

#### Example
This code should run fine:

```python
class MyModel(BaseModel):
    a: Union[A, B, C]

    <some validation of a>

my_instance = MyModel(a=A())
assert isinstance(my_instance.a, C)
```
This code should also run fine for `a=B()` and `a=C()`

### Problem
The type after validation should be reflected in the type-hints for an improved developer experience. Currently it has to be `Union[A, B, C]` to uphold the validation rules of pydantic; however, it can only be `C` post-validation.

#### Fixes
https://github.com/caniko/pydantic-numpy/issues/16

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/usage/exporting_models/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/usage/schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/usage/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/usage/model_config/)
- [X] [Field Types](https://docs.pydantic.dev/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/usage/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/usage/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by Kludex** [comment]
I didn't get it. Do you have a more practical example?

**Comment by caniko** [comment]
There is no practical example, the feature doesn't exist. I updated the description with an example where the feature is needed.

**Comment by sydney-runkle** [comment]
@caniko Perhaps you could just define the type of field `a` as `C`, and add a functional validator to said field to handle coercion from types `A` or `B` to `C`?

**Comment by caniko** [comment]
Thanks @sydney-runkle, that did cross my mind, but it is a half measure and too implicit to reflect the types that are accepted to other developers.

**Comment by sydney-runkle** [comment]
@caniko what about accepting a field `a: Union[A, B, C]` and adding a `computed_field` `a_post` with type `C`?

**Comment by sydney-runkle** [comment]
@caniko I converted this to a discussion, as there's no practical application that this feature request is pertinent to at the moment. Happy to talk more if you want 👍 .

**Comment by maoding** [comment]
Maybe this is a use case: 
I have a pydantic model. e.g. parsing parsing json parameters. Some parameters are optional.
But optional parameters may have default values which are set based on other parameters.

Simplified version:
```python
from pydantic import BaseModel,model_validator
from typing import Self

class Foo(BaseModel):
    a: int
    b: float | None = None


    @model_validator(mode="after")
    def validate_model(self: Self) -> Self:
        if not self.b:
            self.b = 1.0
        return self


foo1 = Foo(a=1)
foo2 = Foo(a=1, b=2.0)

a: float = foo1.b
```

When I run mypy I get an error for the last line. Although I am sure that b is always a float after validation, mypy complains.

Or would you model this differently?
